### ISSUE TITLE
Add rendered_filename argument to derive format from result file suffix

### ISSUE THREAD

**Issue by NAThompson** [xflr6/graphviz#146]
In the examples, we see that `graphviz` adds a suffix to the filename provided based on the format:

```python
>>> dot.render('test-output/round-table.gv', view=True)  # doctest: +SKIP
'test-output/round-table.gv.pdf'
```

Would it not be better to just chose the format based on the filename suffix?

Currently, I have a workaround where I write:

```python
p = Path(filename)
if p.suffix == '.svg':
    dot.format = 'svg'

...

dot.render(p.stem)
```

or else the suffix is appended to the filename twice.

**Comment by xflr6** [comment]
As you see in the example, `filename` specifies the name of the DOT source file to be written. This file is then is rendered by `dot` with `-O` into the source file with an appended `.<format>` (e.g. `.pdf`). So this follows `dot` and delegates the name handling to upstream. +1 that this can be annoying if you don't want the `.gv`-part in your filenames.

The option that comes to my mind is to add a keyword-only optional `rendered_filename` argument and then derive `format` from that (raise if method-argument format and file format differ) and call `dot` specifying the output filename. Can still be used together with `filename` to use different names. WDYT?

**Comment by NAThompson** [comment]
> WDYT?

I think that's a good idea.

**Comment by xflr6** [comment]
>  Is there a way to create the pdf without ever constructing the dot input textfile?

If you need to avoid ever writing the source to disk, you could pipe it and then write the bytes.

```python
In [1]: import pathlib
   ...: import tempfile
   ...: 
   ...: import graphviz
   ...: 
   ...: dot = graphviz.Graph()
   ...: 
   ...: raw = dot.pipe(format='png')  # load the whole file into memory
   ...: 
   ...: with tempfile.NamedTemporaryFile(delete=False) as f:
   ...:     f.write(raw)
   ...:     
   ...: target = pathlib.Path(f.name)
   ...: assert target.stat().st_size
```

> Or at least remove that file from the filesystem after dot operates on it?

Yes, see the `cleanup` argument on [.render()](https://graphviz.readthedocs.io/en/stable/api.html#graphviz.Graph.render).

There is also a proposal to use `tempfile` (similar to above), which we could add as an option.

Let me know in case you want to contribute on any of this.

**Comment by NAThompson** [comment]
> Yes, see the cleanup argument on .render().

Thanks! That's exactly what I needed.

**Comment by xflr6** [comment]
Working on this feature in the main branch.

Maybe take a look at https://graphviz.readthedocs.io/en/latest/changelog.html and https://graphviz.readthedocs.io/en/latest/api.html#graphviz.render and let me know if you have any comments or suggestions.

**Comment by NAThompson** [comment]
This is a major improvement. I do wonder if you could elide the `filename` and `outfile` kwargs and just determine everything from the file suffix.

**Comment by xflr6** [comment]
Glad if it is useful. :)

> I do wonder if you could elide the ``filename`` and ``outfile`` kwargs and just determine everything from the file suffix.

I think you mean something like this, did I get it right?

```python
>>> import graphviz
>>> dot = graphviz.Graph(name='spam')
>>> dot.render(outfile_suffix='.pdf')
'spam.pdf'
```

**Comment by NAThompson** [comment]
I might just do:

```
>>> import graphviz
>>> dot = graphviz.Graph()
>>> dot.render('spam.pdf')
```

I feel like the "mental space" occupied by the `name` conflicts a bit with the (more useful) notion of `title`.. .

**Comment by xflr6** [comment]
Not sure I got it. This will be supported with the new API:

```python
>>> dot.render(outfile='spam.pdf')
```

**Comment by xflr6** [comment]
> I feel like the "mental space" occupied by the ``name`` conflicts a bit with the (more useful) notion of ``title``.. .

Hm, upstream uses `ID` (https://www.graphviz.org/doc/info/lang.html), that would make ``name`` more close, WDYT?

**Comment by NAThompson** [comment]
> Not sure I got it. This will be supported with the new API

Yes, the new API is still a big improvement.

> Hm, upstream uses ID

I would avoid `ID`; seems kinda generic. . .

**Comment by xflr6** [comment]
> Yes, the new API is still a big improvement.

I don't think I got your point yet, can you provide an example maybe?

**Comment by NAThompson** [comment]
Well, if I just do `dot.render('spam.pdf')`, then I don't have to read the docs; it just works. However, I have to look up the docs to get `dot.render(outfile='spam.pdf')`.

**Comment by xflr6** [comment]
> Well, if I just do dot.render('spam.pdf')

Oh, okay now I got it. Well, I would not consider a backwards-incompatible change, which you seem to be suggesting (I assume most users would agree on that). IMHO replicating https://www.graphviz.org/doc/info/command.html#-O by default is not a bad choice for the default

**Comment by NAThompson** [comment]
I think you original suggestion:

```
>>> dot.render(outfile='spam.pdf')
```

is good as the other is backwards incompatible.

**Comment by xflr6** [comment]
That did not get across as intended.

What I meant is adding an ``outfile_suffix`` argument besides ``outfile`` and allow to use either one (raise if both are given and inconsistent), WDYT?

**Comment by NAThompson** [comment]
> What I meant is adding an outfile_suffix argument besides outfile and allow to use either one (raise if both are given and inconsistent), WDYT?

I think that option probably won't be used; bit too complicated IMO. You'd have to read the documentation to know it exists, and nobody reads documentation.

**Comment by NAThompson** [comment]
> Well, I would not consider a backwards-incompatible change, which you seem to be suggesting (I assume most users would agree on that).

I note that the backwards incompatible change is something that surprises nearly everyone. To wit, if I pass `dot.render('foo.pdf')`, it seems extremely unlikely that I want `foo.pdf` to have file type `.gv`.

**Comment by xflr6** [comment]
IMHO such a backwards-incompatible behavioural change (with lots of room for confusion) would be unpythonic.

```
>>> import this
In the face of ambiguity, refuse the temptation to guess.
````

This libary is designed to follow upstream Graphviz when in doubt (e.g. API overlap of ``save()`` and ``render()``).
I am not convinced this is a a bad default in general.

I agree that we can make the API safer to use by making all arguments on the `.render()` method (as oposed to the standalone function) keyword-only. Some deprecation of positional args planned.

**Comment by xflr6** [comment]
Included in [0.19 release](https://pypi.org/project/graphviz/0.19/#files), closing.
